### ISSUE TITLE
BatchStrucutreBuilder

### ISSUE THREAD

**Issue by Osos8128** [jonathanking/sidechainnet#11]
>>> for batch in dataloaders['train']:
....    predicted_angles = model(batch.seqs)
....    sb = scn.BatchedStructureBuilder(batch.int_seqs, predicted_angles)
....    predicted_coords = sb.build()
....    loss = compute_loss(batch.angs, batch.crds,               # True values
....                        predicted_angles, predicted_coords)   # Predicted values

In this part of code i tried to follow the code and your comments but i didn't get. what is the shape of predicted angles BatchedStructureBuilder are expect ? i thought it could be 3 or 12 maybe ?!

**Comment by jonathanking** [comment]
Hi @Osos8128 . The shape expected for BatchStructureBuilder's second argument, `predicted_angles`, is `(Batch_size x L x 12)`. `L` is the length of the proteins in a given batch after being padded to the maximum sequence length for that batch.

**Comment by Osos8128** [comment]
So if my model predicting 3 angles it will be unsupported to use this function ?
and why it is Batch_size x L x 12 ?, although output from pytorch model will be L x batch_size x num_angels ?!

**Comment by jonathanking** [comment]
1. So if my model predicting 3 backbone angles only it will be unsupported to use this function ?
      - Yes, that is correct. BatchedStructureBuilder only supports angle tensors with the shape `(Batch_size x L x 12)`. 
2. and why it is Batch_size x L x 12 ?, although output from pytorch model will be L x batch_size x num_angels ?!
    - This size was chosen because it matched the output of the PyTorch models I have created and used in the past. If you have a Pytorch model that produces `L x batch_size x num_angles` I would recommend simply transposing the matrix, or swapping the two dimensions with (torch.transpose)[https://pytorch.org/docs/master/generated/torch.transpose.html#torch.transpose].
3. Also why you use pad_sequence in models examples although for batch in d['train']: return a batch of proteins with the same length ?!
    - I'm sorry, I am not sure I understand your question. I don't know what you are referring to with `pad_sequence`. If you are using a DataLoader, each DataLoader collates the data so that while a batch contains proteins of various lengths, each protein has been padded to a length that is equal to the length of the largest protein n the batch.
4. And what is the output of this line ? "lengths = sequence.shape[1] - (sequence == 0).all(axis=2).sum(axis=1)".
   - This computes the lengths of all of the proteins in the batch as a `(Batch_size x 1)` sequence of integers.
